File: backend/src/models/Webhook.ts
import mongoose, { Schema, Document } from 'mongoose';
import { IUser } from './User';

export interface IWebhook extends Document {
  url: string;
  secret: string;
  events: string[];
  ownerId: IUser['_id'];
  isActive: boolean;
  lastTriggeredAt?: Date;
  failureCount?: number;
  createdAt: Date;
  updatedAt: Date;
}

const WebhookSchema: Schema = new Schema({
  url: { type: String, required: true },
  secret: { type: String, required: true },
  events: { 
    type: [String], 
    enum: ['payment.success', 'payment.failed', 'subscription.created', 'subscription.expired', 'subscription.cancelled'],
    default: ['payment.success', 'payment.failed']
  },
  ownerId: { type: Schema.Types.ObjectId, ref: 'User', required: true },
  isActive: { type: Boolean, default: true },
  lastTriggeredAt: { type: Date },
  failureCount: { type: Number, default: 0 },
}, { timestamps: true });

WebhookSchema.index({ ownerId: 1 });

export default mongoose.model<IWebhook>('Webhook', WebhookSchema);

File: backend/app/models/webhook.py
from datetime import datetime, timezone
from typing import Optional, Any, List
from beanie import Document, Link, PydanticObjectId
from pydantic import Field
import secrets # Import secrets module for token generation

from app.models.user import User
from app.models.gateway_transaction import GatewayTransaction

class Webhook(Document):
    owner_id: PydanticObjectId = Field(alias="ownerId")
    url: str
    secret: str = Field(default=secrets.token_hex(32)) # Generate secret on creation
    events: List[str] = Field(default=["payment.success", "payment.failed"])
    is_active: bool = Field(default=True, alias="isActive")
    last_triggered_at: Optional[datetime] = Field(default=None, alias="lastTriggeredAt")
    failure_count: int = Field(default=0, alias="failureCount")
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    class Settings:
        name = "webhooks"
        indexes = [
            "owner_id",
            "url",
            "is_active",
            "events",
        ]
